webapp/Dockerfile

```diff
@@ -0,0 +1,26 @@
+ARG NODE_VERSION=16.14.2
+ARG ALPINE_VERSION=3.15
+ARG NGINX_VERSION=1.21.6
+
+FROM node:${NODE_VERSION}-alpine${ALPINE_VERSION} AS development
+ENV NODE_ENV=development
+RUN apk add --no-cache dumb-init
```

Comment:
Just to make image lighter I'd have fewer layers, try to group directives. It is less legible but when deploying it helps to save space

Reply:
I think very often (at least last time I was using docker) the layers will break, and a full removal of the image will be required to rebuild anyway. So hard to leverage on caching.
Theoretically is better, but I'm not sure this works practically.

Reply:
I will do some tests to see exactly how much space we save by grouping the directives versus how many seconds we do by caching the dependencies each time we rebuild the images.

In my opinion, time > disk space, but I agree that it's hard to justify without more information.

Reply:
I've done some tests with these changes:
<details>
<summary>Click to see changes</summary>

```diff
diff --git a/transcendence-app/Dockerfile b/transcendence-app/Dockerfile
index bd686cc..13b3963 100644
--- a/transcendence-app/Dockerfile
+++ b/transcendence-app/Dockerfile
@@ -3,22 +3,18 @@ ARG ALPINE_VERSION=3.15

 FROM node:${NODE_VERSION}-alpine${ALPINE_VERSION} AS development
 ENV NODE_ENV=development
-RUN apk add --no-cache dumb-init
 WORKDIR  /transcendence-app
-COPY ["package.json", "package-lock.json", "./"]
-RUN npm clean-install
 COPY . .
-RUN npm run build
+RUN apk add --no-cache dumb-init && npm clean-install && npm run build
 ENTRYPOINT ["/usr/bin/dumb-init", "--"]
 CMD ["npm", "run", "start:dev"]

 FROM node:${NODE_VERSION}-alpine${ALPINE_VERSION} AS production
 ENV NODE_ENV=production
-RUN apk add --no-cache dumb-init
 WORKDIR  /transcendence-app
 COPY ["package.json", "package-lock.json", "./"]
-RUN npm clean-install --production
 COPY --from=development --chown=node:node /transcendence-app/dist ./dist
+RUN apk add --no-cache dumb-init && npm clean-install --production
 USER node
 ENTRYPOINT ["/usr/bin/dumb-init", "--"]
 CMD ["node", "dist/main"]

```

```diff
diff --git a/webapp/Dockerfile b/webapp/Dockerfile
index e8f4b31..3f20d83 100644
--- a/webapp/Dockerfile
+++ b/webapp/Dockerfile
@@ -4,11 +4,9 @@ ARG NGINX_VERSION=1.21.6

 FROM node:${NODE_VERSION}-alpine${ALPINE_VERSION} AS development
 ENV NODE_ENV=development
-RUN apk add --no-cache dumb-init
 WORKDIR /webapp
-COPY ["package.json", "package-lock.json", "./"]
-RUN npm clean-install
 COPY . .
+RUN apk add --no-cache dumb-init && npm clean-install
 ENTRYPOINT ["/usr/bin/dumb-init", "--"]
 CMD ["npm", "start"]

@@ -17,10 +15,8 @@ ENV NODE_ENV=production
 ARG REACT_APP_NAME
 ENV REACT_APP_NAME=${REACT_APP_NAME}
 WORKDIR /webapp
-COPY ["package.json", "package-lock.json", "./"]
-RUN npm clean-install --production
 COPY . .
-RUN npm run build
+RUN npm clean-install --production && npm run build

 FROM nginx:${NGINX_VERSION}-alpine AS production
 COPY --from=builder --chown=nginx:nginx /webapp/build /usr/share/nginx/html

```

</details>

Optimizing disk space by grouping directives saves 2MB of disk space.
<details>
<summary>Click to see results</summary>

Image size development - with disk size optimization
```
REPOSITORY                           TAG                  IMAGE ID       CREATED              SIZE
ft_transcendence_webapp              latest               3526b74e2744   32 seconds ago       369MB
ft_transcendence_transcendence-app   latest               9146b88eb919   About a minute ago   409MB

```

Image size development - with caching optimization
```
REPOSITORY                           TAG                  IMAGE ID       CREATED              SIZE
ft_transcendence_webapp              latest               39583615cf11   53 seconds ago       370MB
ft_transcendence_transcendence-app   latest               4988fa1fcb8e   About a minute ago   410MB
```

Image size production - with disk size optimization
```
REPOSITORY                           TAG                  IMAGE ID       CREATED              SIZE
ft_transcendence_webapp              latest               b91eab7312c4   14 seconds ago       24MB
ft_transcendence_transcendence-app   latest               66c7eba73357   About a minute ago   126MB
```

Image size production - with caching optimization
```
REPOSITORY                           TAG                  IMAGE ID       CREATED              SIZE
ft_transcendence_webapp              latest               408d3bf353eb   4 days ago           24MB
ft_transcendence_transcendence-app   latest               d4af61464114   2 days ago           126MB
```

</details>

On my machine, for development, `npm clean-install` inside Docker takes around 25 seconds for transcendence-app and 35 seconds for webapp, without caching. With caching, it takes 0 seconds.
Optimizing build time by using layer caching saves 60 seconds of build time.
<details>

```
% docker-compose  up -d --build
...
Building transcendence-app
...
 => [development 5/7] RUN npm clean-install        26.4s 
...

...
Building webapp
...
 => [development 5/6] RUN npm clean-install        36.6s
...

```

</details>

Reply:
For me is reasonable to keep layers then since we are not effectively saving anything